index.js

```diff
@@ -1,153 +1,224 @@
-  }
-  compilation.fileDependencies.push(templateFile);
-  return fs.readFileAsync(templateFile, 'utf8')
-    // If the file could not be read log a error
-    .catch(function() {
-      return Promise.reject(new Error('HtmlWebpackPlugin: Unable to read HTML template "' + templateFile + '"'));
-    });
+  });
 };
 
-/*
- * Compile the html template and push the result to the compilation assets
+/**
+ * Evaluates the child compilation result
+ * Returns a promise
  */
-HtmlWebpackPlugin.prototype.emitHtml = function(compilation, htmlTemplateContent, templateParams, outputFilename) {
-  var html;
-  // blueimp-tmpl processing
-  try {
-    html = tmpl(htmlTemplateContent, templateParams);
-  } catch(e) {
-    return Promise.reject(new Error('HtmlWebpackPlugin: template error ' + e));
-  }
-
-  // Inject link and script elements into an existing html file
-  if (this.options.inject) {
-    html = this.injectAssetsIntoHtml(html, templateParams);
+HtmlWebpackPlugin.prototype.evaluateCompilationResult = function(compilationResult) {
+  if(!compilationResult) {
+    return Promise.reject('The child compilation didn\'t provide a result');
   }
-
-  // Minify the html output
-  if (this.options.minify) {
-    var minify = require('html-minifier').minify;
-    // If `options.minify` is set to true use the default minify options
-    var minifyOptions = _.isObject(this.options.minify) ? this.options.minify : {};
-    html = minify(html, minifyOptions);
+  // Strip the leading 'var '
+  var source = compilationResult.source().substr(3);
+  // Evaluate code and cast to string
+  var newSource;
+  try {
+    newSource = vm.runInThisContext(source);
+  } catch (e) {
+    return Promise.reject(e);
   }
+  return typeof newSource === 'string' ?
+    Promise.resolve(newSource) :
+    Promise.reject('The loader "' + this.options.template + '" didn\'t return html.');
+};
 
-  compilation.assets[outputFilename] = {
-    source: function() {
-      return html;
-    },
-    size: function() {
-      return html.length;
-    }
-  };
+/**
+ * Html post processing
+ *
+ * Returns a promise
+ */
+HtmlWebpackPlugin.prototype.postProcessHtml = function(html, chunks, compilation) {
+  var self = this;
+  var webpackStatsJson = compilation.getStats().toJson();
+  var assets = self.htmlWebpackPluginAssets(compilation, webpackStatsJson, chunks);
+  return Promise.resolve()
+    // Favicon
+    .then(function() {
+      if (self.options.favicon) {
+        return self.addFileToAssets(self.options.favicon, compilation)
+          .then(function(faviconBasename){
+            assets.favicon = faviconBasename;
+          });
+      }
+    })
+    // Template processing
+    .then(function() {
+      var templateParams = {
+        webpack: webpackStatsJson,
+        webpackConfig: compilation.options,
+        htmlWebpackPlugin: {
+          files: assets,
+          options: self.options,
+        }
+      };
+      if (self.options.compile === true) {
```

Comment:
hey @bebraw thanks for the feedback.. 
Actually I have a new (hopefully) better idea.
Instead of using a template processing part I will use a blueimp template loader by default.
This will reduce the complexity and will still support blueimp templates